flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java

```diff
@@ -2025,6 +2047,123 @@ private void freeNoLongerUsedSlots(AllocatedSlotReport allocatedSlotReport) {
+            AllocationID allocationId,
+            ResourceProfile resourceProfile) {
+        if (!allocationsDirectory.exists() && !allocationsDirectory.mkdirs()) {
+            log.debug(
+                    "Allocations directory {} doesn't exist and cannot be created.",
+                    allocationsDirectory.toPath());
+            return;
+        }
+
+        // Let's try to write the slot allocations on file
+        final File slotAllocationSnapshotFile = slotAllocationFile(slotId.getSlotNumber());
+        try (ObjectOutputStream oos =
+                new ObjectOutputStream(new FileOutputStream(slotAllocationSnapshotFile))) {
+            oos.writeObject(
+                    new LocalSlotAllocationSnapshot(
+                            slotId, jobId, jobTargetAddress, allocationId, resourceProfile));
+            log.debug(
+                    "Successfully written allocation state metadata file {} for job {} and allocation {}.",
+                    slotAllocationSnapshotFile.toPath(),
+                    jobId,
+                    allocationId);
+        } catch (IOException e) {
+            log.debug(
+                    "Cannot write the local allocations state. File {} for job {} and allocation {}.",
+                    slotAllocationSnapshotFile.toPath(),
+                    jobId,
+                    allocationId,
+                    e);
+        }
+    }
+
+    private void tryCleanupSlotAllocationSnapshot(int requestedIndex) {
+        if (!allocationsDirectory.exists()) {
+            log.debug(
+                    "There is no local allocations snapshot directory to cleanup {}.",
+                    allocationsDirectory.toPath());
+            return;
+        }
+
+        // Let's try to write the slot allocations on file
+        final File slotAllocationSnapshotFile = slotAllocationFile(requestedIndex);
+        try {
+            FileUtils.deleteFileOrDirectory(slotAllocationSnapshotFile);
+            log.debug(
+                    "Successfully deleted allocation state metadata file {}.",
+                    slotAllocationSnapshotFile.toPath());
+        } catch (IOException e) {
+            log.debug(
+                    "Cannot delete the local allocations state file {}.",
+                    slotAllocationSnapshotFile.toPath(),
+                    e);
+        }
+    }
+
+    private File slotAllocationFile(int slotIndex) {
+        return new File(allocationsDirectory.getAbsolutePath(), slotIndex + ".bin");
+    }
+
+    /**
+     * This method tries to repopulate the {@link JobTable} and {@link TaskSlotTable} from the local
+     * filesystem in a best-effort manner
+     */
+    private void tryLoadLocalAllocationSnapshots() {
+        if (!allocationsDirectory.exists()) {
+            log.debug(
+                    "There is no local allocations snapshot directory to load from {}.",
+                    allocationsDirectory.toPath());
+            return;
+        }
```

Comment:
Same here. If we create the directory at the start of `TaskExecutor` then we don't need this change here.